gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java

```diff
@@ -960,49 +962,76 @@ private void checkQuota(DagNode<JobExecutionPlan> dagNode) throws IOException {
       String proxyUser = ConfigUtils.getString(dagNode.getValue().getJobSpec().getConfig(), AzkabanProjectConfig.USER_TO_PROXY, null);
       String specExecutorUri = DagManagerUtils.getSpecExecutorUri(dagNode);
       boolean proxyUserCheck = true;
+      int proxyQuotaIncrement;
+      Map<String, Integer> usersQuotaIncrement = new HashMap<>();
+      StringBuilder requesterMessage = new StringBuilder();
+
       if (proxyUser != null) {
-        proxyUserCheck = incrementMapAndCheckQuota(proxyUserToJobCount, proxyUser, dagNode);
+        proxyQuotaIncrement = incrementJobCountAndCheckUserQuota(proxyUserToJobCount, proxyUser, dagNode);
+        proxyUserCheck = proxyQuotaIncrement < 0;  // proxy user quota check failed
+        if (!proxyUserCheck) {
+          requesterMessage.append(String.format(
+              "Quota exceeded for proxy user %s on executor %s : quota=%s, runningJobs=%d%n",
+              proxyUser, specExecutorUri, getQuotaForUser(proxyUser),
+              proxyUserToJobCount.get(DagManagerUtils.getUserQuotaKey(proxyUser, dagNode))));
+        }
       }
 
       String serializedRequesters = DagManagerUtils.getSerializedRequesterList(dagNode);
       boolean requesterCheck = true;
-      String requesterMessage = null;
+
       if (serializedRequesters != null) {
         for (ServiceRequester requester : RequesterService.deserialize(serializedRequesters)) {
-          requesterCheck &= incrementMapAndCheckQuota(requesterToJobCount, requester.getName(), dagNode);
-          if (!requesterCheck && requesterMessage == null) {
-            requesterMessage = "Quota exceeded for requester " + requester.getName() + " on executor " + specExecutorUri + ": quota="
-                + getQuotaForUser(requester.getName()) + ", runningJobs=" + requesterToJobCount.get(DagManagerUtils.getUserQuotaKey(requester.getName(), dagNode));
+          int userQuotaIncrement = incrementJobCountAndCheckUserQuota(requesterToJobCount, requester.getName(), dagNode);
+          boolean thisRequesterCheck;
+          thisRequesterCheck = userQuotaIncrement < 0;  // user quota check failed
+          usersQuotaIncrement.put(requester.getName(), Math.max(userQuotaIncrement, 0));
+          requesterCheck = requesterCheck && thisRequesterCheck;
+          if (!thisRequesterCheck) {
+            requesterMessage.append(String.format(
+                "Quota exceeded for requester %s on executor %s : quota=%s, runningJobs=%d%n",
+                requester.getName(), specExecutorUri, getQuotaForUser(requester.getName()),
```

Comment:
given multi-threading, `getQuotaForUser` may differ from what it was during `incrementJobCountAndCheck...`; need to return then-current value from that method.